src/clp_logging/handlers.py

```diff
@@ -792,3 +800,198 @@ def __init__(
+        )
+        self.last_timestamp_ms: int = floor(time.time() * 1000)  # convert to ms and truncate
+        self.ostream.write(
+            FourByteEncoder.encode_preamble(
+                self.last_timestamp_ms, self.timestamp_format, self.timezone
+            )
+        )
+
+    def __init__(
+        self,
+        s3_bucket: str,
+        stream: Optional[IO[bytes]] = None,
+        enable_compression: bool = True,
+        timestamp_format: Optional[str] = None,
+        timezone: Optional[str] = None,
+        aws_access_key_id: Optional[str] = None,
+        aws_secret_access_key: Optional[str] = None,
+        part_limit: Optional[int] = None
+    ) -> None:
+        super().__init__()
+        self.closed: bool = False
+        self.enable_compression: bool = enable_compression
+        self.local_buffer: io.BytesIO = io.BytesIO()
+        if stream is None:
+            stream = self.local_buffer
+        self.ostream: IO[bytes] = stream
+        self.timestamp_format: str
+        self.timezone: str
+        self.timestamp_format, self.timezone = _init_timeinfo(timestamp_format, timezone)
+        self.init(stream)
+
+        self.s3_bucket: str = s3_bucket
+        self.remote_folder_path: Optional[str] = None
+        self.remote_file_count: int = 0
+        self.start_timestamp: datetime = datetime.datetime.now()
+        self.obj_key: str = self._remote_log_naming(self.start_timestamp)
+        self.s3_resource: boto3.resources.factory.s3.ServiceResource = boto3.resource("s3")
+        try:
+            self.s3_client = boto3.client(
+                "s3",
+                aws_access_key_id=aws_access_key_id,
+                aws_secret_access_key=aws_secret_access_key
+            ) if aws_access_key_id and aws_secret_access_key else boto3.client("s3")
+        except botocore.exceptions.NoCredentialsError:
+            raise RuntimeError("AWS credentials not found. Please configure your credentials.")
+        except botocore.exceptions.ClientError as e:
+            raise RuntimeError(f"Failed to initialize AWS client: {e}")
+        self.buffer_size: int = 1024 * 1024 * 0.5
+        self.part_limit: int = part_limit if part_limit else 10000
+        self.uploaded_parts: List[Dict[str, int | str]] = []
+        self.upload_index: int = 1
+        create_ret: Dict[str, Any] = self.s3_client.create_multipart_upload(
+            Bucket=self.s3_bucket, Key=self.obj_key, ChecksumAlgorithm="SHA256"
+        )
+        self.upload_id: int = create_ret["UploadId"]
+        if not self.upload_id or not isinstance(self.upload_id, str):
+            raise RuntimeError("Failed to obtain a valid Upload ID from S3.")
+
+
+    def _remote_log_naming(self, timestamp: datetime.datetime) -> str:
+        self.remote_folder_path: str = f"logs/{timestamp.year}/{timestamp.month}/{timestamp.day}"
+
+        new_filename: str
+        upload_time: str = timestamp.strftime("%Y-%m-%d-%H%M%S")
```

Comment:
It's probably better to put a raw Unix timestamp instead, since the year, month, day are already present in the key (as the directories names).